cogeo_mosaic/backends/dynamodb.py

```diff
@@ -231,16 +268,55 @@ def get_assets(self, x: int, y: int, z: int) -> List[str]:
 
     def _fetch_dynamodb(self, quadkey: str) -> Dict:
         try:
-            return self.table.get_item(Key={"quadkey": quadkey}).get("Item", {})
+            return self.table.get_item(
+                Key={"mosaicId": self.mosaic_name, "quadkey": quadkey}
+            ).get("Item", {})
         except ClientError as e:
             status_code = e.response["ResponseMetadata"]["HTTPStatusCode"]
             exc = _HTTP_EXCEPTIONS.get(status_code, MosaicError)
             raise exc(e.response["Error"]["Message"]) from e
 
-    def _table_exist(self) -> bool:
+    def _table_exists(self) -> bool:
         """Check if the Table already exists."""
         try:
             _ = self.table.table_status
             return True
         except self.table.meta.client.exceptions.ResourceNotFoundException:
             return False
+
+    def _mosaic_exists(self) -> bool:
+        """Check if the mosaic already exists in the Table."""
+        item = self.table.get_item(
+            Key={"mosaicId": self.mosaic_name, "quadkey": "-1"}
+        ).get("Item", {})
+
+        return True if item else False
+
+    def clean(self):
+        """clean MosaicID from dynamoDB Table."""
+        logger.debug(f"Deleting items for mosaic {self.mosaic_name}...")
+
+        # get items
+        resp = self.table.query(
+            KeyConditionExpression="#mosaicId = :mosaicId",
+            ExpressionAttributeNames={"#mosaicId": "mosaicId", "#quadKey": "quadKey"},
+            ExpressionAttributeValues={":mosaicId": {"S": self.mosaic_name}},
+            ProjectionExpression="#quadKey",
+        )
```

Comment:
this isn't working

Reply:
IMO this isn't great since it requires a full table scan any time you want to delete records.

Reply:
This doesn't quite do a full table scan, since it queries for a partition key (`mosaicId`) meaning that it only searches within a single partition. I agree that deleting records in a loop isn't the most efficient, @vincentsarago  mentioned that the operation would be not be performed often enough to affect costs/performance

Reply:
That being said @vincentsarago, why isn't it working?

Reply:
I'll try to reproduce and post the error!

Reply:
> it queries for a partition key (`mosaicId`) meaning that it only searches within a single partition

Thanks for clearing that up, it makes a lot of sense! I was worried that a very large table with many `mosaicid`s would be excessively slow, but in effect it's pretty similar to the operation that happens currently when you delete a table.

Reply:
💯

Reply:
```
ClientError                               Traceback (most recent call last)
<ipython-input-7-dc3e028485c1> in <module>
----> 1 m.clean()

~/Workspace/DevSeed/cogeo-mosaic/cogeo_mosaic/backends/dynamodb.py in clean(self)
    302             ExpressionAttributeNames={"#mosaicId": "mosaicId", "#quadKey": "quadKey"},
    303             ExpressionAttributeValues={":mosaicId": {"S": self.mosaic_name}},
--> 304             ProjectionExpression="#quadKey",
    305         )
    306 

~/Workspace/venv/py37/lib/python3.7/site-packages/boto3/resources/factory.py in do_action(self, *args, **kwargs)
    518             # instance via ``self``.
    519             def do_action(self, *args, **kwargs):
--> 520                 response = action(self, *args, **kwargs)
    521 
    522                 if hasattr(self, 'load'):

~/Workspace/venv/py37/lib/python3.7/site-packages/boto3/resources/action.py in __call__(self, parent, *args, **kwargs)
     81                     operation_name, params)
     82 
---> 83         response = getattr(parent.meta.client, operation_name)(**params)
     84 
     85         logger.debug('Response: %r', response)

~/Workspace/venv/py37/lib/python3.7/site-packages/botocore/client.py in _api_call(self, *args, **kwargs)
    314                     "%s() only accepts keyword arguments." % py_operation_name)
    315             # The "self" in this scope is referring to the BaseClient.
--> 316             return self._make_api_call(operation_name, kwargs)
    317 
    318         _api_call.__name__ = str(py_operation_name)

~/Workspace/venv/py37/lib/python3.7/site-packages/botocore/client.py in _make_api_call(self, operation_name, api_params)
    624             error_code = parsed_response.get("Error", {}).get("Code")
    625             error_class = self.exceptions.from_code(error_code)
--> 626             raise error_class(parsed_response, operation_name)
    627         else:
    628             return parsed_response

ClientError: An error occurred (ValidationException) when calling the Query operation: One or more parameter values were invalid: Condition parameter type does not match schema type
```

Reply:
🤦 `quadKey` -> `quadkey`